### PR TITLE
Generalize Runtime and provide JSON handler

### PR DESCRIPTION
**Pull request by amarrella**
This PR generalizes the runtime to support other values than JSON, we (or downstream dependencies) can create new handlers for other formats: Dhall, Binary, XML, whatever...

I also refactored the code so that lambdas don't need anymore to handle the runtime client logic (see the `examples`), but can directly use the JSON handler (or other handlers) simplifying downstream logic. 

The modules for the "Raw" client are still exposed and usable by downstream dependencies that need that level of control.

### CODE REVIEW

**Comment by amarrella** [comment]
> 👍 Seems reasonable, perhaps an example of something non-JSON would be also nice then?

@andreyk0 : Thanks, good point, this also made me realize decode needs to be effectful, otherwise some decoders  (e.g. dhall's, which performs side effects to resolve imports) will fail. Will add the dhall decoder as an example and make decoding effectful

**Comment by amarrella** [comment]
@andreyk0: thanks for the feedback! This is done now, for dhall

**Comment by amarrella** [comment]
Note: I still have to handle output well. Assuming `Show` is probably not the best option. I think I'll need to add an "encode" parameter to the runtime for maximum flexibility. This way it will also be possible to write hybrid handlers (with input and output both strongly typed, but different)

**Comment by amarrella** [comment]
This is complete now :) I revised the approach to use typeclasses as suggested by @barambani .
Now it successfully decodes and encodes both JSON and Dhall values.

**Comment by amarrella** [comment]
Actually, looks like in the "real" lambda world input needs to be valid json, so at the very least i need to escape the text. Still wip

File: README.md [hunk 1]
```diff
@@ -55,9 +54,9 @@ data User = User {
   name :: Text, 
   accountId :: Natural
 } deriving (Generic, Show, FromJSON, ToJSON)
+  deriving LambdaDecode via (LambdaFromJSON User)
```
Comment (barambani) [inline README.md]: nice

**Comment by amarrella** [comment]
Sorry, I keep pushing to this branch because I want a clean master. Last commit makes the handler function effectful, otherwise the runtime would be very limited in what it can achieve :)

**Comment by amarrella** [comment]
> Recalling our discussion, is it the handle operation or the encoding that can be effectful ? In which phase the dhal file needs to be resolved ?

Actually both. The dhall is effectful in the [decoding](https://hackage.haskell.org/package/dhall-1.26.1/docs/Dhall.html#v:input), but the lambda itself also needs to "do something" when it's not a pure data transformation (e.g. trigger a step function)

**Comment by barambani** [comment]
Right, makes sense.

**Comment by amarrella** [comment]
I'm unsure whether I should use MonadError as in https://github.com/EarnestResearch/aws-lambda-haskell-runtime-client/pull/36/commits/1b5131b02b2fe2aa077dfa572c5936a846286f8e or I should instead give the caller access to the client via `MonadReader`. 

If I give them access via MonadReader, I also need to remove the logic of posting responses and errors from the handler and hand that over to the client, otherwise we risk that the caller calls `postError` and the handler does a `postResponse` afterwards. Any advice?

File: examples/api-gateway/Main.hs [hunk 2]
```diff
@@ -1,24 +1,16 @@
...
-  forever $ echo client
+main = runStderrLoggingT $ handler echo
 
-echo :: (MonadLogger m, MonadIO m) => RuntimeClient APIGatewayInputEvent APIGatewayOutputEvent m -> m ()
-echo RuntimeClient{..} = do
-  Event{..} <- getNextEvent
-  case eventBody of
-    Right ie -> postResponse eventID $ apiGatewayOutputEvent $ ie ^. body
-    Left e   -> postError eventID $ Error "Unexpected Error" $ pack e
+echo :: (LambdaEncode e, MonadIO m, MonadLogger m, MonadError e m) => APIGatewayInputEvent -> m APIGatewayOutputEvent
```
Comment (barambani) [inline examples/api-gateway/Main.hs]: The least powerful structure that we need in the examples is `Applicative`
```suggestion
echo :: Applicative m => APIGatewayInputEvent -> m APIGatewayOutputEvent
```

**Comment by amarrella** [comment]
Can you re-approve if you are happy with the changes?

**Pull request merged**